sdk/keyvault/azure-keyvault-keys/azure/keyvault/keys/crypto/_models.py

```diff
@@ -2,11 +2,176 @@
+        """
+        key = self._key_client.get_key(self._key_name)
+        return int.from_bytes(key.key.n, "big").bit_length()  # type: ignore[attr-defined]
+
+    def public_key(self) -> RSAPublicKey:
+        """The `RSAPublicKey` associated with this private key.
+
+        :returns: The `RSAPublicKey` associated with the key.
+        :rtype: RSAPublicKey
+        """
+        key = self._key_client.get_key(self._key_name)
+        e = int.from_bytes(key.key.e, "big")  # type: ignore[attr-defined]
+        n = int.from_bytes(key.key.n, "big")  # type: ignore[attr-defined]
+        public_numbers = RSAPublicNumbers(e, n)
+        return public_numbers.public_key()
+
+    def sign(
+        self,
+        data: bytes,
+        padding: AsymmetricPadding,
+        algorithm: Union[Prehashed, HashAlgorithm],
+    ) -> bytes:
+        """Signs the data.
+
+        :param bytes data: The data to sign, as bytes.
+        :param padding: The padding to use. Supported paddings are `PKCS1v15` and `PSS`. For `PSS`, the only supported
+            mask generation function is `MGF1`. See https://learn.microsoft.com/azure/key-vault/keys/about-keys-details
+            for details.
+        :type padding: AsymmetricPadding
+        :param algorithm: The algorithm to sign with. Only `HashAlgorithm`s are supported -- specifically, `SHA256`,
+            `SHA384`, and `SHA512`.
+        :type algorithm: Prehashed or HashAlgorithm
+
+        :returns: The signature, as bytes.
+        :rtype: bytes
+        """
+        if isinstance(algorithm, Prehashed):
+            raise ValueError("`Prehashed` algorithms are unsupported. Please provide a `HashAlgorithm` instead.")
+
+        mapped_algorithm = SIGN_ALGORITHM_MAP.get(type(algorithm))
+        if mapped_algorithm is None:
+            raise ValueError(f"Unsupported algorithm: {algorithm.name}")
+
+        # If PSS padding is requested, use the PSS equivalent algorithm
+        if isinstance(padding, PSS):
+            mapped_algorithm = PSS_MAP.get(type(mapped_algorithm))
+
+            # Public mgf property was only added in https://github.com/pyca/cryptography/pull/9582
+            # _mgf property has been available in every version of the PSS class, so we use it as a backup
+            try:
+                mgf = padding.mgf
+            except AttributeError:
+                mgf = padding._mgf
+            if not isinstance(mgf, MGF1):
+                raise ValueError(f"Unsupported MGF: {mgf}")
+
+        # The only other padding accepted is PKCS1v15
+        elif not isinstance(padding, PKCS1v15):
+            raise ValueError(f"Unsupported padding: {padding.name}")
+
+        digest = Hash(algorithm)
+        digest.update(data)
+        result = self._crypto_client.sign(cast(SignatureAlgorithm, mapped_algorithm), digest.finalize())
+        return result.signature
+
+    def private_numbers(self) -> RSAPrivateNumbers:  # pylint:disable=docstring-missing-return,docstring-missing-rtype
```

Comment:
You can actually type scenarios that will always raise:
```suggestion
    def private_numbers(self) -> NoReturn:
```
Though I'm not sure if this will have warnings with deviation from the base protocol...

Reply:
I went ahead and implemented `private_numbers`, but applied this feedback to `private_bytes`. Mypy and pylint seem fine with it, based on a quick local check